File: backend/utils/ticker_utils.py
"""
Ticker Formatting Utilities
Provides consistent ticker formatting across the application
"""

import logging
from typing import Optional
import re

logger = logging.getLogger("ticker_utils")

# Ticker validation pattern: 1-10 alphanumeric characters, optional .EXCHANGE suffix
TICKER_PATTERN = re.compile(r'^[A-Z0-9]{1,10}(\.[A-Z]{2,10})?$', re.IGNORECASE)


def format_ticker_for_eodhd(ticker: str, exchange_code: Optional[str] = None) -> str:
    """
    Format ticker with exchange suffix for EODHD API.

    EODHD API requires tickers in format: SYMBOL.EXCHANGE
    Examples: AAPL.US, BMW.XETRA, 7203.T

    Args:
        ticker: Base ticker symbol (e.g., 'AAPL' or 'AAPL.US')
        exchange_code: Exchange code (e.g., 'US', 'XETRA', 'T')
                      If None and ticker has no exchange, defaults to 'US'

    Returns:
        Formatted ticker with exchange suffix (e.g., 'AAPL.US')

    Examples:
        >>> format_ticker_for_eodhd('AAPL', 'US')
        'AAPL.US'

        >>> format_ticker_for_eodhd('AAPL.US', None)
        'AAPL.US'

        >>> format_ticker_for_eodhd('BMW', 'XETRA')
        'BMW.XETRA'

        >>> format_ticker_for_eodhd('AAPL')
        'AAPL.US'  # Default to US
    """
    # Normalize to uppercase
    ticker = ticker.upper().strip()

    # If ticker already has exchange suffix, return as-is
    if "." in ticker:
        return ticker

    # If exchange code provided, append it
    if exchange_code:
        exchange_code = exchange_code.upper().strip()
        return f"{ticker}.{exchange_code}"

    # Default to US exchange for backward compatibility
    logger.debug(f"No exchange specified for {ticker}, defaulting to .US")
    return f"{ticker}.US"


def format_ticker_for_company(company) -> str:
    """
    Format ticker with exchange suffix using Company model.

    Args:
        company: Company model instance with exchange relationship

    Returns:
        Formatted ticker (e.g., 'AAPL.US')

    Raises:
        ValueError: If company has no exchange and no default available
    """
    # Use company's exchange if available
    if company.exchange:
        return format_ticker_for_eodhd(
            ticker=company.ticker,
            exchange_code=company.exchange.code
        )

    # Fallback to just ticker (will use default .US in format_ticker_for_eodhd)
    logger.warning(
        f"Company {company.ticker} (ID: {company.id}) has no exchange assigned, "
        f"defaulting to .US suffix"
    )
    return format_ticker_for_eodhd(ticker=company.ticker)


def validate_ticker_format(ticker: str) -> bool:
    """
    Validate ticker format.

    Valid formats:
    - AAPL
    - AAPL.US
    - BMW.XETRA
    - 7203.T

    Args:
        ticker: Ticker string to validate

    Returns:
        True if valid, False otherwise

    Examples:
        >>> validate_ticker_format('AAPL')
        True

        >>> validate_ticker_format('AAPL.US')
        True

        >>> validate_ticker_format('invalid!!!')
        False
    """
    return bool(TICKER_PATTERN.match(ticker))


def parse_ticker(ticker: str) -> dict:
    """
    Parse ticker into symbol and exchange components.

    Args:
        ticker: Ticker string (e.g., 'AAPL.US' or 'AAPL')

    Returns:
        Dictionary with 'symbol' and 'exchange' keys

    Examples:
        >>> parse_ticker('AAPL.US')
        {'symbol': 'AAPL', 'exchange': 'US'}

        >>> parse_ticker('AAPL')
        {'symbol': 'AAPL', 'exchange': None}
    """
    ticker = ticker.upper().strip()

    if "." in ticker:
        symbol, exchange = ticker.split(".", 1)
        return {"symbol": symbol, "exchange": exchange}
    else:
        return {"symbol": ticker, "exchange": None}


def get_bare_ticker(ticker: str) -> str:
    """
    Extract bare ticker symbol without exchange suffix.

    Args:
        ticker: Ticker with or without exchange (e.g., 'AAPL.US' or 'AAPL')

    Returns:
        Bare ticker symbol (e.g., 'AAPL')

    Examples:
        >>> get_bare_ticker('AAPL.US')
        'AAPL'

        >>> get_bare_ticker('AAPL')
        'AAPL'
    """
    parsed = parse_ticker(ticker)
    return parsed["symbol"]

File: frontend/utils/tickerUtils.ts
/**
 * Ticker Formatting Utilities
 *
 * Provides consistent ticker formatting across the application.
 * Ensures compliance with EODHD API standards.
 */

/**
 * Asset type classification
 */
export type AssetType = 'stock' | 'etf' | 'forex' | 'crypto' | 'unknown';

/**
 * Parsed ticker information
 */
export interface ParsedTicker {
  symbol: string;           // Base symbol (e.g., 'AAPL', 'EURUSD')
  exchange: string | null;  // Exchange code (e.g., 'US', 'FOREX')
  fullTicker: string;       // Complete ticker with exchange (e.g., 'AAPL.US')
  displayTicker: string;    // Ticker for UI display
  assetType: AssetType;     // Classified asset type
}

/**
 * Format ticker for EODHD API calls
 * Ensures ticker has proper exchange suffix
 *
 * @param ticker - Raw ticker input
 * @param defaultExchange - Default exchange if none specified (default: 'US')
 * @returns Formatted ticker with exchange suffix (e.g., 'AAPL.US', 'EURUSD.FOREX')
 */
export function formatTickerForAPI(ticker: string, defaultExchange: string = 'US'): string {
  if (!ticker) return '';

  const normalized = ticker.toUpperCase().trim();

  // Already has exchange suffix
  if (normalized.includes('.')) {
    return normalized;
  }

  // Auto-detect forex pairs (6 characters, common pattern)
  // Examples: EURUSD, GBPJPY, USDJPY
  if (normalized.length === 6 && /^[A-Z]{6}$/.test(normalized)) {
    const commonCurrencies = ['USD', 'EUR', 'GBP', 'JPY', 'AUD', 'CAD', 'CHF', 'NZD', 'CNY', 'HKD', 'SGD', 'SEK', 'NOK', 'DKK'];
    const first = normalized.substring(0, 3);
    const second = normalized.substring(3, 6);

    if (commonCurrencies.includes(first) && commonCurrencies.includes(second)) {
      return `${normalized}.FOREX`;
    }
  }

  // Default to provided exchange
  return `${normalized}.${defaultExchange}`;
}

/**
 * Get display ticker for UI
 * Strategy: ALWAYS show exchange suffix for clarity and consistency
 *
 * @param ticker - Ticker with or without exchange
 * @returns Formatted ticker for display (always with exchange suffix)
 */
export function getDisplayTicker(ticker: string): string {
  if (!ticker) return '';

  // Ensure ticker has exchange suffix
  const fullTicker = formatTickerForAPI(ticker);

  // Return with exchange suffix for consistency
  return fullTicker;
}

/**
 * Get bare ticker symbol without exchange suffix
 * Use this ONLY when interacting with third-party services that don't support exchange suffixes
 * (e.g., TradingView)
 *
 * @param ticker - Ticker with or without exchange
 * @returns Bare symbol without exchange (e.g., 'AAPL' from 'AAPL.US')
 */
export function getBareTicker(ticker: string): string {
  if (!ticker) return '';

  const normalized = ticker.toUpperCase().trim();

  if (normalized.includes('.')) {
    return normalized.split('.')[0];
  }

  return normalized;
}

/**
 * Parse ticker into components
 *
 * @param ticker - Ticker to parse
 * @returns Parsed ticker information
 */
export function parseTicker(ticker: string): ParsedTicker {
  if (!ticker) {
    return {
      symbol: '',
      exchange: null,
      fullTicker: '',
      displayTicker: '',
      assetType: 'unknown'
    };
  }

  const normalized = ticker.toUpperCase().trim();
  let symbol: string;
  let exchange: string | null;

  if (normalized.includes('.')) {
    [symbol, exchange] = normalized.split('.');
  } else {
    symbol = normalized;
    exchange = null;
  }

  const fullTicker = formatTickerForAPI(ticker);
  const assetType = classifyAssetType(fullTicker);

  return {
    symbol,
    exchange,
    fullTicker,
    displayTicker: fullTicker,  // Always show with exchange
    assetType
  };
}

/**
 * Classify asset type from ticker
 *
 * @param ticker - Full ticker with exchange
 * @returns Asset type classification
 */
export function classifyAssetType(ticker: string): AssetType {
  if (!ticker) return 'unknown';

  const normalized = ticker.toUpperCase();

  // Forex detection
  if (normalized.includes('.FOREX') || normalized.includes('.FX')) {
    return 'forex';
  }

  // Crypto detection
  if (normalized.includes('.CC') || normalized.includes('BTC') || normalized.includes('ETH')) {
    return 'crypto';
  }

  // Check if symbol is 6-char forex pair
  const symbol = normalized.split('.')[0];
  if (symbol.length === 6 && /^[A-Z]{6}$/.test(symbol)) {
    const commonCurrencies = ['USD', 'EUR', 'GBP', 'JPY', 'AUD', 'CAD', 'CHF', 'NZD', 'CNY', 'HKD', 'SGD', 'SEK', 'NOK', 'DKK'];
    const first = symbol.substring(0, 3);
    const second = symbol.substring(3, 6);

    if (commonCurrencies.includes(first) && commonCurrencies.includes(second)) {
      return 'forex';
    }
  }

  // ETF detection (common ETF tickers)
  const etfTickers = ['SPY', 'QQQ', 'VOO', 'VTI', 'IWM', 'DIA', 'EEM', 'GLD', 'SLV', 'TLT', 'IEF', 'LQD', 'HYG'];
  if (etfTickers.includes(symbol)) {
    return 'etf';
  }

  // Default to stock
  return 'stock';
}

/**
 * Get asset type badge color for UI
 *
 * @param assetType - Asset type
 * @returns Tailwind CSS color classes
 */
export function getAssetTypeBadgeColor(assetType: AssetType): string {
  switch (assetType) {
    case 'stock':
      return 'bg-blue-500 text-white';
    case 'etf':
      return 'bg-purple-500 text-white';
    case 'forex':
      return 'bg-yellow-500 text-gray-900';
    case 'crypto':
      return 'bg-orange-500 text-white';
    default:
      return 'bg-gray-500 text-white';
  }
}

/**
 * Validate ticker format
 *
 * @param ticker - Ticker to validate
 * @returns True if valid ticker format
 */
export function isValidTicker(ticker: string): boolean {
  if (!ticker) return false;

  // Pattern: 1-10 alphanumeric characters, optional .EXCHANGE suffix
  const pattern = /^[A-Z0-9]{1,10}(\.[A-Z]{2,10})?$/i;
  return pattern.test(ticker.trim());
}

/**
 * Sanitize ticker input by removing dangerous characters
 *
 * @param ticker - Raw ticker input
 * @returns Sanitized ticker (alphanumeric + dot only)
 */
export function sanitizeTicker(ticker: string): string {
  if (!ticker) return '';
  return ticker.toUpperCase().replace(/[^A-Z0-9.]/g, '');
}
